codegen/src/androidMain/kotlin/com/freeletics/khonshu/codegen/internal/ActivityComponentLookup.kt

```diff
@@ -1,33 +1,46 @@
 package com.freeletics.khonshu.codegen.internal
 
 import android.content.Context
-import android.os.Bundle
+import androidx.activity.ComponentActivity
 import androidx.compose.runtime.DisallowComposableCalls
+import androidx.compose.runtime.ProvidableCompositionLocal
+import androidx.compose.runtime.staticCompositionLocalOf
 import androidx.lifecycle.SavedStateHandle
 import androidx.lifecycle.SavedStateViewModelFactory
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.ViewModelProvider
-import androidx.lifecycle.ViewModelStoreOwner
 import java.io.Closeable
 import kotlin.reflect.KClass
 
 @InternalCodegenApi
-public inline fun <reified C : Any, P : Any> component(
-    viewModelStoreOwner: ViewModelStoreOwner,
-    context: Context,
-    parentScope: KClass<*>,
-    arguments: Bundle?,
-    crossinline factory: @DisallowComposableCalls (P, SavedStateHandle, Bundle) -> C,
+public interface ActivityComponentProvider {
+    public fun <T> provide(scope: KClass<*>): T
+}
+
+@InternalCodegenApi
+public val LocalActivityComponentProvider: ProvidableCompositionLocal<ActivityComponentProvider> =
+    staticCompositionLocalOf {
+        throw IllegalStateException("ActivityComponentProvider was not provided by Activity")
+    }
+
+@InternalCodegenApi
+public inline fun <C : Any, reified AC : Any, P : Any> component(
+    activity: ComponentActivity,
+    requestedScope: KClass<*>,
+    activityScope: KClass<*>,
+    activityParentScope: KClass<*>,
+    crossinline factory: @DisallowComposableCalls (P, SavedStateHandle) -> AC,
 ): C {
-    val store = ViewModelProvider(
-        viewModelStoreOwner,
-        SavedStateViewModelFactory(),
-    )[ActivityComponentViewModel::class.java]
-    return store.getOrCreate(C::class) {
-        val parentComponent = context.findComponentByScope<P>(parentScope)
-        val savedStateHandle = store.savedStateHandle
-        factory(parentComponent, savedStateHandle, arguments ?: Bundle.EMPTY)
+    if (requestedScope != activityScope) {
+        return activity.findComponentByScope(requestedScope)
```

Comment:
This allows to use the provider to also look up other components higher up in the hierarchy, e.g. AppScope. That way the parent lookup doesn't need to know whether the parent scope is an Activity scope or not.